Tutorial examples

kuniss opened the pull request:
@linkrope please have a first look on the tutorial examples. 
I did not move them to a _tutorial_ sub directory yet, as discussed.

Only the _decla-appl.eag_ example has been changed for didactic reasons - to differentiate meta and hyper non-terminals uniquely.
All other examples are new and quite trivial. 

May be the decl-appl test input is superfluous. But it was handy during the presentation...

kuniss committed:
Added 2 examples used at the strumenta talk.

kuniss committed:
Use "i" instead of pipe for counting

to be conform to the presentation.

kuniss committed:
Added BNF kind decl-appl

kuniss committed:
Added 2 simple examples for EBNF operators.

kuniss committed:
Made PLUS or MINUS mandatory in the target language.

kuniss committed:
Added comment about reference in tutorial

kuniss committed:
Corrected indentation.

kuniss committed:
Renamed hyper symbol 'id' to 'ident' for tutorial

to let the student easier differentiate between the hyper and
the meta symbol. Otherwise this would be hard for a beginner.

linkrope commented on test/decl-appl/test1:
```diff
@@ -0,0 +1,5 @@
+DECL ab
+DECL aba
```
Why not "ab" and "ba"?
First thought is that "ab" is a prefix of "aba", which is the wrong track.

kuniss replied:
Hm, just flew out of my fingers. 
"ba" will work too.

kuniss replied:
... Followed your suggestion using "ab" and "ba".

linkrope commented on test/decl-appl/test3:
```diff
@@ -0,0 +1,5 @@
+APPL aba
```
A minimal example would be

    APPL ab

    DECL ab

However, a second identifier could help showing the wrong order and the right order side by side:

    DECL ab

    APPL ab
    APPL ba

    DECL ba

BTW: without the requirement for an LEAG, we could even change the EAG to allow application before declaration. Wouldn't this be nice?

kuniss replied:
Looks like a better example!
Will play with it in regard to the tutorial text.

Regarding your "BTW": Let's talk offline.

kuniss replied:
Followed your suggestion for having right and wrong order side by side.
Additionally, moved the test3 content to test2 as test2 is not used in the tutorial at, and removed superfluous test3.

kuniss replied:
> BTW: without the requirement for an LEAG, we could even change the EAG to allow application before declaration. Wouldn't this be nice?

Isn't this covered by the chapter [Changing the validation semantic of the example language](https://github.com/linkrope/gamma/wiki/Table-Lookup-and-Semantic-Validation#changing-the-validation-semantic-of-the-example-language)?

linkrope commented on example/number.eag:
```diff
@@ -0,0 +1,25 @@
+// A simple example referenced in the tutorial for showing how the optional EBNF operator works
```
`Digits` vs. `digits` takes so much space of the example, and the issue, that the digit sequence can be empty is distracting. Moreover, `[ '+' | '-' ]` is the right way to define a sign, while `( '+' | '-')` would be the wrong way.

How about simple expressions?

    Expr: ['+' | '-'] Term.
    Term: 'x'.

vs.

    Expr: Term ('+' | '-') Term.
    Term: 'x'

kuniss replied:
Yes, maybe a better example.
I have to think about and check against the tutorial text.

kuniss replied:
Not worked on yet, until commit https://github.com/linkrope/gamma/pull/10/commits/c71be10dd828adcd430a3eb710845b86b4fe0d33

kuniss replied:
@linkrope 

What do you think about the following two example specifications?
For option operator:
```
Expr<+ Sign>: <Sign>[ <+'Plus':Sign> '+' | <+'Minus':Sign> '-' ]<+'Plus':Sign> Term.
Sign = 'Plus' | 'Minus'.
Term: 'x'.
```

For grouping operator:
```
ExprG<+ Op>: Term <Op>( <+'Addition':Op> '+' | <+'Subtraction':Op> '-' ) Term.
Op = 'Addition' | 'Subtraction'.
Term: 'x'.
```

Or is it better to use `Sign = 'Plus' | 'Minus'.` in both examples for simplicity? What do you think?

I would name the examples, _ebnf-option.eag_ and _ebnf-grouping.eag_ as this are really minimal examples and quite meaningless otherwise.

linkrope replied:
Even if it's "+" or "-" in both cases, it's better to distinguish between "sign" and "op":
one next addition for "op" could be "OR", which does not make sense for "sign";
and one possbility for "sign" could be <empty>, which does not make sense for "op".

"op" and "sign" will go different ways in more realisitic examples, so we should keep them separate from the beginning.

BTW: replacing the missing sign with "+" only works with numbers, but not with booleans or strings. The best option for the sign would be to put the term into `[...]`, and to get it out unchanged, if there is no sign: `<- Term :Term, + Term: Expr>`. However, this would not work wit left-to-right evaluation. The second best option for the sign is the additional empty alternative. A "sign" is "+" or "-" or not there at all.

kuniss replied:
> Even if it's "+" or "-" in both cases, it's better to distinguish between "sign" and "op":
one next addition for "op" could be "OR", which does not make sense for "sign";
and one possbility for "sign" could be , which does not make sense for "op".
>"op" and "sign" will go different ways in more realisitic examples, so we should keep them separate from the beginning.

Added the 2 simple, meaningless examples, one with "sign", one with "op", both lower case.

linkrope commented on example/abc-bnf.eag:
```diff
@@ -0,0 +1,20 @@
+// a^n b^n c^n -> |^n in BNF 
```
I would like to put the examples without EBNF into a separate directory.

How about `example/bnf/abc.eag`?
Then, we can even drop the suffix "-bnf".
And we can add `example/bnf/README.md`: "only for illustrative purposes, don't do this"

My problem with "bnf" is, that it sounds like a cool new feature:
the boring stuff is under `example/`, while the next big thing is under `example/bnf/`.

How about `example/no-ebnf`?

kuniss replied:
Good idea!
What about `example/bnf-legacy`? Legacy is an odd word for SW engineers. :smile:

linkrope replied:
Don't like "legacy" in this context.
I thought about "straight" or "simple", but don't like them either.

I still like the term "lowering" for the transformation of an EBNF into a BNF. As a starting point, however, "lowering" wouzld be as bad as "lowering". So, just èxample/bnf/`?

kuniss replied:
Yes, let's go with `example/bnf`. We even consider the reader to be familiar with EBNF and BNF concepts. So, he/she should know that EBNF notation is more concise and so has advantages over BNF.

kuniss replied:
I have moved BNF like grammars to example/bnf and renamed all EBN-like grammars to have no "-ebnf" suffix.

linkrope commented on example/abc-bnf.eag:
```diff
@@ -0,0 +1,20 @@
+C<- "i" N: N>: "c" C <N>.
+C<- : N>: .
+
+// build generator: dub build
```
We should not mix grammars with commands.
(This was the Oberon way to do it.)
It's too easy to forget these commands when there is a breaking change.

Better point to the test implementations.
As soon as a reader understands the simple test implementations, they have lots of running examples available.

kuniss replied:
What do you mean with "point to the test implementation"? Do not get it.

linkrope replied:
Just like this: "https://github.com/linkrope/gamma/blob/master/test/slag.d#L5"

Or better mention the test implementation under `test/` and that a reader can find lots of examples how to call the generator as well as the generated compilers there.

kuniss replied:
Will remove it. Was some sort of redundancy for the strumenta talk to avoid switching back and forth between slides and gitpod.

kuniss replied:
Removed all build instructions from EAG tutorial files.

linkrope commented on example/decl-appl-bnf.eag:
```diff
@@ -0,0 +1,38 @@
+x<+ "a": x>: "a".
+x<+ "b": x>: "b".
+
+ident*<+ id>:
```
Do you really need this change?
One level down, is's meta `x` as well as hyper `x`.

Better add a sentence to the tutorial that such a "lifting" is very common for "terminal symbols" (non-terminals with `*`).

kuniss replied:
For newbies which are not familiar with concepts and which part of specification is representing which concept it is very hard to follow a describing text if there are equal named identifiers with different meanings.

It also gets very hard to reference them in the tutorial text right: As concepts are not clear yet for a newbie, the reference "meta-nonterminal 'id'" or "hyper-nonterminal 'id'" could not be differentiated and the tutorial would need to talk about positions of 'id'.

That's at the end the real reason for having different names for meta and hyper non-terminals. 
I'm looking from a didactic point of view.
Talking about "lifting" would not help a lot here as it is just another concept needing explanation...

linkrope replied:
BTW: it's `ident` / `ident` for Oberon-0, but `id` / `id` for Decl/Appl to make better use of the space on a slide. I thought about `xs` instead of your `id`, but `xs '';' Table` destroys the intuition. Why `xs`? Because the left recursion for `ident` only makes sense if you have letters and digits, but only letters at the beginning. But with only "a" and "b", there is no motivation for `identTail`.

Also from an educational perspective, I don't like "ident" vs. "id". Even as experts, we have no idea whether it should be `id <ident>` or the other way around `ident <id>`. And shall we put the `id` into the table, or the `ident`? Naming both `ident` is the best way to tell: "don't care!" You want to find the (hyper) identifier from APPL in the (meta) table.

(BTW: you somehow avoided with `x` / `x`` in your text by just talking about meta `x`.)

If you're still not convinced: a compromise could be `Variable <ident>`.

kuniss replied:
Referencing it as _affix variable_ `<ident>` in the text is a good idea!

May be it is also a good idea to talk about lifting. 
I remember,  we maybe want to introduce a syntactic shortcut for that...

kuniss replied:
Renamed "ident" back to "id" at examples/bnf/decl-apl.eag and examples/decl-appl.eag (where I also renamed it, previously). Reworked the tutorial text (not in this commit, see wiki).

linkrope reviewed (changes requested):
Two general remarks:
1. even the diff does not like "no newline at end of file", where red symbols are shown
2. we should definitely have a grammar style guide

Indentation is sometimes 4, sometimes 8.
I'm not sure in which situation the period should be on its own line.
But it should be indented then?
Are you aware of (EBNF) grammar style guidelines, which we can copy as a start?
I would even reformat the existing examples to match this style.

kuniss committed:
Added trailing emty line

to get rid of complains (diff editor).

kuniss committed:
Moved ab.eag to example/bnf and renamed ab-ebnf.eag

using the now free name ab.eag.

kuniss committed:
Moved *-bnf.eag to example/bnf and renamed.

kuniss committed:
Removed build instruction comments

originially helpful for the instrucmenta talk.

kuniss committed:
Corrected/Added reference to EBNF-like example

kuniss committed:
Reworked decla-appl source examples

to avoid readers get on the wrong track while seeing a prefix.
This resolves the first part the discussion
https://github.com/linkrope/gamma/pull/10#pullrequestreview-1008983532

kuniss committed:
Showing right order and wrong order side by side.

Resolves discussion review discussion
https://github.com/linkrope/gamma/pull/10#discussion_r899048651

kuniss committed:
Reordered APPL statements in decl-appl example

in test1 to get  it prepared as predecessor for test3
for better description flow in the tutorial.

kuniss committed:
Moved content of decl-appl test3 to test2, removed test3

as original test2 is not used at all in the tutorial.

kuniss committed:
Renamed hype rnon-terminal ident to id

as discussed in the review discussion
https://github.com/linkrope/gamma/pull/10#discussion_r899129325

kuniss committed:
Corrected example tests after renaming of ab-ebnf.eag.

kuniss commented:
@linkrope, I need help on the [check-style build issue...](https://github.com/linkrope/gamma/runs/8052967717?check_suite_focus=true)
After having rebased my PR branch to gamma/master all your changes are in. Only my changes remains. And still the check step fails.  Unfortunately, I have no clue.

I reworked a lot of the tutorial examples according to our discussion here. Only the discussion point https://github.com/linkrope/gamma/pull/10#discussion_r899074914 regarding example/number.eag remains.
Could you please push the `Resolve conversion` button after review for the resolved discussion to have a better overview.
I also updated the Wiki, mainly https://github.com/linkrope/gamma/wiki/Table-Lookup-and-Semantic-Validation to reflect the changes. However, the Gitpod hands-on parts still needs to get repeated and validated according to the repetition result.

linkrope commented:
> @linkrope, I need help on the [check-style build issue...](https://github.com/linkrope/gamma/runs/8052967717?check_suite_focus=true) After having rebased my PR branch to gamma/master all your changes are in. Only my changes remains. And still the check step fails. Unfortunately, I have no clue.

The following change is required to avoid the issue with building the check tool: [Update dscanner](https://github.com/linkrope/gamma/commit/1fd88ce19f6f385757b56f19420dac4307a5868a)
Just rebase once again.

linkrope commented on example/ab.eag:
```diff
@@ -1,7 +1,7 @@
-// a^n b^n  -> i^n  without EBNF
+// a^n b^n  -> i^n  with EBNF
```
`example/bnf` collects the special grammar files.
The idea is that "with EBNF" is the normal way to do it.

So I would rather remove "with EBNF" here.
(It could be added to each grammar file, so it's better to not add it at all.)

kuniss replied:
Done.

linkrope commented on example/bnf/ab.eag:
```diff
@@ -0,0 +1,8 @@
+N = 'i' N | .
+
+S: <+ N: N> A <N> B <N>.
+A: <+ 'i' N: N> 'a' A <N> | <+ : N> .
```
Should be as close as possible to `example/bnf/abc.eag`:
don't even use the shortcut for alternatives of `A`.

(And add blank line between rules fo `A` and rules for `B`.)

kuniss replied:
Done.

linkrope commented on example/bnf/decl-appl.eag:
```diff
@@ -0,0 +1,33 @@
+        
+idTail: 
+        <- id, + id1> x<x> idTail<id x, id1>
+    |   <- id, + id>
```
Apply consistent style: don't use `|` in any hyper-rule in `example/bnf`´.
Because of this, indent by 4 spaces instead of 8.

kuniss replied:
Done.

linkrope commented on example/bnf/expr.eag:
```diff
@@ -0,0 +1,33 @@
+    | <+ "1": Number> "1".
+    
+    
+// ExprTail shorter:
```
In `example/bnf` don't use any shortcut: not even `|` for hyper-rules.

kuniss replied:
Done. Also removed the commented out EBNF specification alternatives as they can be examined in the according EBNF example.

linkrope commented on example/decl-appl.eag:
```diff
@@ -5,11 +5,11 @@ Table = | id ';' Table.
     <, Table>
         { <- Table, + Table1>
-            'DECL' id< id>
+            'DECL' id<id>
```
It's always a space before `<`.

kuniss replied:
Done. 
You meant, there should be a space between each hyper non-terminal and its affix parameters list starting with an `<`, right?

kuniss committed:
Adjusted examples as discussed at the PR #10

kuniss committed:
Added simple examples for EBNF operators.
